File: supabase/functions/mail-ingest/handlers.ts
import {
  IngestMeta,
  ParsedTable,
  ParseDebug,
  buildRowObject,
  buildSkipErrorText,
  isValidRunDate,
  jsonResponse,
  normalizeAttachmentType,
} from './shared.ts'
import { parseAttachment } from './parse.ts'
import {
  getSupabaseAdmin,
  insertFilePayloadRecord,
  insertFileRecord,
  markPipelineRunAfterReset,
  refreshPipelineRunAfterIngest,
} from './supabase.ts'

type AttachmentParseResult = {
  table: ParsedTable | null
  debug: ParseDebug
  parseFailureMessage: string | null
}

async function parseFile(
  attachmentType: 'xlsx' | 'csv',
  file: File,
): Promise<{ bytes: Uint8Array; parsed: AttachmentParseResult }> {
  const bytes = new Uint8Array(await file.arrayBuffer())
  let parsedResult: { table: ParsedTable | null; debug: ParseDebug } | null = null
  let parseFailureMessage: string | null = null

  try {
    parsedResult = parseAttachment(attachmentType, bytes)
  } catch (error) {
    parseFailureMessage =
      error instanceof Error ? error.message : 'Unknown parser error'
  }

  return {
    bytes,
    parsed: {
      table: parsedResult ? parsedResult.table : null,
      debug: parsedResult
        ? parsedResult.debug
        : {
            type: attachmentType,
            summary: { error: 'parse_failed' },
          },
      parseFailureMessage,
    },
  }
}

function resolveTableOutcome(parsed: AttachmentParseResult) {
  const table = parsed.table
  const status = table
    ? 'ingested'
    : parsed.parseFailureMessage
      ? 'error'
      : 'skipped'
  const header = table ? table.header : []
  const rowCount = table ? table.rows.length : 0
  const errorText = table
    ? null
    : parsed.parseFailureMessage || buildSkipErrorText('no_valid_table_block', parsed.debug)

  return {
    table,
    status,
    header,
    rowCount,
    errorText,
  }
}

function validateIngestMeta(meta: IngestMeta) {
  return (
    meta.action === 'ingest' &&
    isValidRunDate(meta.run_date) &&
    !!String(meta.primary_topic || '').trim() &&
    ['primary', 'secondary'].includes(String(meta.topic_role || ''))
  )
}

export async function handleReset(req: Request) {
  const payload = await req.json()
  const runDate = payload?.run_date

  if (payload?.action !== 'reset' || !isValidRunDate(runDate)) {
    return jsonResponse({ ok: false, error: 'Invalid reset payload' }, 400)
  }

  const supabase = getSupabaseAdmin()
  const { error } = await supabase.from('ingest_files').delete().eq('run_date', runDate)

  if (error) {
    throw error
  }

  await markPipelineRunAfterReset(supabase, runDate)

  return jsonResponse({
    ok: true,
    action: 'reset',
    run_date: runDate,
  })
}

export async function handleIngest(req: Request) {
  const form = await req.formData()
  const metaRaw = form.get('meta')
  const file = form.get('file')

  if (typeof metaRaw !== 'string' || !(file instanceof File)) {
    return jsonResponse({ ok: false, error: 'Missing multipart meta or file' }, 400)
  }

  let meta: IngestMeta
  try {
    meta = JSON.parse(metaRaw)
  } catch (_error) {
    return jsonResponse({ ok: false, error: 'Invalid meta JSON' }, 400)
  }

  if (!validateIngestMeta(meta)) {
    return jsonResponse({ ok: false, error: 'Invalid ingest payload' }, 400)
  }

  const attachmentType = normalizeAttachmentType(meta.attachment_type, file.name, file.type)
  if (!attachmentType) {
    return jsonResponse({ ok: false, error: 'Unsupported attachment type' }, 415)
  }

  const supabase = getSupabaseAdmin()
  let fileId: string | null = null

  try {
    const { bytes, parsed } = await parseFile(attachmentType, file)
    const outcome = resolveTableOutcome(parsed)

    fileId = await insertFileRecord(
      supabase,
      meta,
      attachmentType,
      outcome.status,
      outcome.header,
      outcome.rowCount,
      outcome.errorText,
    )
    await insertFilePayloadRecord(supabase, fileId, file.type, bytes)

    if (outcome.table && outcome.table.rows.length) {
      const rowPayload = outcome.table.rows.map((row, index) => ({
        file_id: fileId,
        run_date: meta.run_date,
        row_index: index + 1,
        row_json: buildRowObject(outcome.table!.header, row),
      }))

      const { error: rowsError } = await supabase.from('ingest_rows').insert(rowPayload)
      if (rowsError) {
        await supabase
          .from('ingest_files')
          .update({ status: 'error', error_text: rowsError.message })
          .eq('id', fileId)
        throw rowsError
      }
    }

    await refreshPipelineRunAfterIngest(supabase, meta.run_date)

    return jsonResponse({
      ok: true,
      status: outcome.status,
      file_id: fileId,
      rows: outcome.rowCount,
      debug: parsed.debug,
    })
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown ingest error'

    if (fileId) {
      await supabase
        .from('ingest_files')
        .update({ status: 'error', error_text: message })
        .eq('id', fileId)
      await refreshPipelineRunAfterIngest(supabase, meta.run_date)
    } else {
      await insertFileRecord(
        supabase,
        meta,
        attachmentType,
        'error',
        [],
        0,
        message,
      )
      await refreshPipelineRunAfterIngest(supabase, meta.run_date)
    }

    return jsonResponse({
      ok: false,
      error: message,
    }, 500)
  }
}

File: ingest_service/handlers.py
from __future__ import annotations

from dataclasses import dataclass

from fastapi import UploadFile

from ingest_service.models import ResetPayload
from ingest_service.parse import parse_attachment
from ingest_service.storage import (
    insert_file_payload_record,
    insert_file_record,
    insert_row_records,
    mark_pipeline_run_after_reset,
    refresh_pipeline_run_after_ingest,
)


@dataclass(frozen=True)
class RuntimeHandlers:
    reset_handler: object
    ingest_handler: object


def normalize_attachment_type(meta_type: object, filename: str, content_type: str) -> str | None:
    lowered_type = str(meta_type or "").lower()
    lowered_name = str(filename or "").lower()
    lowered_content_type = str(content_type or "").lower()

    if (
        lowered_type == "xlsx"
        or lowered_name.endswith(".xlsx")
        or "openxmlformats-officedocument.spreadsheetml.sheet" in lowered_content_type
    ):
        return "xlsx"
    if lowered_type == "csv" or lowered_name.endswith(".csv") or "csv" in lowered_content_type:
        return "csv"
    return None


def create_runtime_handlers(connection) -> RuntimeHandlers:
    def reset_handler(payload: ResetPayload):
        mark_pipeline_run_after_reset(connection, payload.run_date)
        return {"ok": True, "action": payload.action, "run_date": payload.run_date}

    def ingest_handler(meta: dict[str, object], upload: UploadFile):
        attachment_type = normalize_attachment_type(
            meta.get("attachment_type"),
            upload.filename or str(meta.get("attachment_name") or ""),
            upload.content_type or "",
        )
        if not attachment_type:
            raise ValueError("Unsupported attachment type")

        payload = upload.file.read()
        parsed = parse_attachment(attachment_type, payload)
        status = "ingested" if parsed.table else "skipped"
        header = parsed.table.header if parsed.table else []
        rows = parsed.table.rows if parsed.table else []
        file_id = insert_file_record(
            connection,
            meta={
                "run_date": str(meta.get("run_date") or ""),
                "message_id": str(meta.get("message_id") or ""),
                "thread_id": str(meta.get("thread_id") or ""),
                "message_date": str(meta.get("message_date") or ""),
                "message_subject": str(meta.get("message_subject") or ""),
                "primary_topic": str(meta.get("primary_topic") or ""),
                "matched_topic": str(meta.get("matched_topic") or ""),
                "topic_role": str(meta.get("topic_role") or ""),
                "attachment_name": str(meta.get("attachment_name") or upload.filename or ""),
            },
            attachment_type=attachment_type,
            status=status,
            header=header,
            row_count=len(rows),
            error_text=None,
        )
        insert_file_payload_record(
            connection,
            file_id=file_id,
            file_content_type=upload.content_type,
            bytes_payload=payload,
        )
        if rows:
            insert_row_records(
                connection,
                file_id=file_id,
                run_date=str(meta.get("run_date") or ""),
                rows=[
                    {
                        str(header[index]): str(row[index] if index < len(row) else "")
                        for index in range(len(header))
                        if str(header[index]).strip()
                    }
                    for row in rows
                ],
            )
        refresh_pipeline_run_after_ingest(connection, str(meta.get("run_date") or ""))
        return {
            "ok": True,
            "status": status,
            "file_id": file_id,
            "rows": len(rows),
            "debug": {
                "type": parsed.debug.type,
                "summary": parsed.debug.summary,
            },
        }

    return RuntimeHandlers(reset_handler=reset_handler, ingest_handler=ingest_handler)
